### DISCUSSION TITLE
Personal Bio Auth Example

### DISCUSSION THREAD

**Discussion by drteresavasquez** [nss-evening-cohort-14/evening-client-side#242] [Class Notes and Discussion]
Personal Bio Auth Example

```javascript
import React, { useState, useEffect } from 'react';
import firebase from 'firebase/app';  // you may get an error from firebase. if so, delete /app
import 'firebase/auth'; // you may get an error from firebase. if so, delete this line
import { useHistory } from 'react-router-dom';

function App() {
  const [admin, setAdmin] = useState(null);

  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
      } else if (admin || admin === null) {
        setAdmin(false);
      }
    });
  }, []);
  
  return(...);
}

export default App;
```

**Comment by drteresavasquez** [comment]
The `.env` file is NOT magic or wizardy. It is literally a file for you to store keys and values. The same way we use it with the firebase config file. We have talked about this at length before. How are your API keys being used in the config file? This means that you can store any values in the file as long as they follow the `REACT_APP_YOUR_KEY_NAME=XXX` convention and call the values in your code using `process.env.REACT_APP_YOUR_KEY_NAME`

You can get the authentication uid from firebase under authentication and users OR you can login and log the value. Either way, the values will be the same. My example for auth looks at the value from the `.env` file and compares it with the uid for the authenticated user. If they match the user is admin.

**Comment by tsekeres** [comment]
Trying to trace the purpose of the having auth code in three different places. 
1. index.js seems to confirm admin or "authed" user and sets admin to true if appropriate.
2. auth.js in the past two projects has defined what the onClick will do. (Sign in or Sign out user)
3. auth.js in your example seems to be rechecking if authed user is indeed present and again sets admin to true.
4. So now I am trying to figure out what should go at the top of my Routes.js in the PrivateRoute code that again checks is admin is true and decides which components to let the user see.

A. is my logic on track?
B. why does the auth.js seem redundant?

**Reply by drteresavasquez** [reply]
`auth.js` has the functions for logging in an out. And, while it does play into this, this example is only checking for the user's current authenticated status.

You are right on points 1 & 2, but on point 3, `auth.js` is not being used in this example so that is not the case.

In routes, what are you checking for? Likely if the user is an admin or not, which is what you said in your post, so that they can see routes, right, So when you call `<Routes />` you also need to pass it the value of admin, which will either be true or false. So, `<Routes admin={admin} />`. Everything else is the same. Your logic may need to change in your private routes, but you should be able to play around with that logic comparison to identify if a user has access to a private route.

Based on the true or false, you can determine private routes and where the user can navigate.

**Reply by tsekeres** [reply]
mhmm, mhmm. So the code above is a suggestion for index.js. And auth.js is left out because you chose not to have all users sign in and out. But instead only care if it is the admin account user.

**Reply by drteresavasquez** [reply]
This code has always gone in App/index.js when using auth because it's watching for the user state to change and we need to pass that to its children. When a users auth status updates this runs.

**Reply by tsekeres** [reply]
Thank you. 
Another question, I know you'll tell me if it should be a ticket:

I also don't know where the signInWithPopup should be. I don't have that anywhere now that auth is empty and doesn't have signInUser anymore.

**Reply by drteresavasquez** [reply]
Place it wherever you think it's needed. If you want info from other classmates about where they put theirs, yes create a ticket 😊 (or just ask them)

**Reply by drteresavasquez** [reply]
Also, why is auth.js empty? You still need those functions. Once you put those back, you can determine where you want sign in to be. You've done this on every app that requires sign in so it's not different than that unless you want it to be.

**Reply by tsekeres** [reply]
I guess I'm confused on how it happens without a button. Auth has only had buttons that control when things happen. I will now move onto the next level in Blooms and apply some knowledge. Thanks

**Reply by drteresavasquez** [reply]
You can use a button if you want. In the walk through I created a login route. You can put a button in there if you'd like or another route you create. Technically since you're ensuring that only you can access private routes you can put it wherever seems natural for you as well. 

Remember that button or link is a reusable component and you can place it anywhere you'd like and set the onClick to call either the login/logout function.

**Comment by cmeffley** [comment]
I noticed that my Google UID is different in each of my Firebase Projects. Is that a Google thing or is something weird on my end? Just curious about this, since as of now,  my admin log in is working *knocks on wood*

**Reply by drteresavasquez** [reply]
Each app has a different key. It's a different app so one app doesn't know about the other and it's a random, unique identifier specific to each of your apps so that's why those are all different.

**Reply by cmeffley** [reply]
Makes sense, thanks!
